x-pack/plugins/endpoint/public/embeddables/resolver/models/indexed_process_tree.ts

```diff
@@ -15,21 +15,90 @@ import { levelOrder as baseLevelOrder } from '../lib/tree_sequencers';
   const idToValue = new Map<string, ResolverEvent>();
+  const idToAdjacent = new Map<string, AdjacentProcessMap>();
+
+  function emptyAdjacencyMap(id: string): AdjacentProcessMap {
+    return {
+      self: id,
+      parent: null,
+      firstChild: null,
+      previousSibling: null,
+      nextSibling: null,
+      level: 1,
+    };
+  }
+
+  const roots: ResolverEvent[] = [];
 
   for (const process of processes) {
-    idToValue.set(uniquePidForProcess(process), process);
+    const uniqueProcessPid = uniquePidForProcess(process);
+    idToValue.set(uniqueProcessPid, process);
+
+    const currentProcessAdjacencyMap: AdjacentProcessMap =
+      idToAdjacent.get(uniqueProcessPid) || emptyAdjacencyMap(uniqueProcessPid);
+    idToAdjacent.set(uniqueProcessPid, currentProcessAdjacencyMap);
+
     const uniqueParentPid = uniqueParentPidForProcess(process);
-    const processChildren = idToChildren.get(uniqueParentPid);
-    if (processChildren) {
-      processChildren.push(process);
+    const currentProcessSiblings = idToChildren.get(uniqueParentPid);
+
+    if (currentProcessSiblings) {
+      const previousProcessId = uniquePidForProcess(
+        currentProcessSiblings[currentProcessSiblings.length - 1]
+      );
+      currentProcessSiblings.push(process);
+      /**
+       * Update adjacency maps for current and previous entries
+       */
+      idToAdjacent.get(previousProcessId)!.nextSibling = uniqueProcessPid;
+      currentProcessAdjacencyMap.previousSibling = previousProcessId;
+      if (uniqueParentPid) {
+        currentProcessAdjacencyMap.parent = uniqueParentPid;
+      }
     } else {
       idToChildren.set(uniqueParentPid, [process]);
+
+      if (uniqueParentPid) {
+        /**
+         * Get the parent's map, otherwise set an empty one
+         */
+        const parentAdjacencyMap =
+          idToAdjacent.get(uniqueParentPid) ||
+          (idToAdjacent.set(uniqueParentPid, emptyAdjacencyMap(uniqueParentPid)),
+          idToAdjacent.get(uniqueParentPid))!;
+        // set firstChild for parent
+        parentAdjacencyMap.firstChild = uniqueProcessPid;
+        // set parent for current
+        currentProcessAdjacencyMap.parent = uniqueParentPid || null;
+      } else {
+        // In this case (no unique parent id), it must be a root
+        roots.push(process);
+      }
     }
   }
 
+  /**
+   * Scan adjacency maps from the top down and assign levels
+   */
+  function traverseLevels(currentProcessMap: AdjacentProcessMap, level: number = 1): unknown {
+    const nextLevel = level + 1;
+    if (currentProcessMap.nextSibling) {
+      traverseLevels(idToAdjacent.get(currentProcessMap.nextSibling)!, level);
+    }
+    if (currentProcessMap.firstChild) {
+      traverseLevels(idToAdjacent.get(currentProcessMap.firstChild)!, nextLevel);
+    }
+    currentProcessMap.level = level;
+    return void 0;
+  }
+
+  for (const treeRoot of roots) {
+    traverseLevels(idToAdjacent.get(uniquePidForProcess(treeRoot))!);
```

Comment:
@oatkiller This is what I ended up replacing the level getter with. Should only touch each node in the tree once.